libraries/chain/market_evaluator.cpp

```diff
@@ -241,13 +250,14 @@ void_result call_order_update_evaluator::do_apply(const call_order_update_operat
       old_collateralization = call_obj->collateralization();
 
       d.modify( *call_obj, [&]( call_order_object& call ){
-          call.collateral += o.delta_collateral.amount;
-          call.debt       += o.delta_debt.amount;
-          if( call.debt > 0 )
-          {
-             call.call_price  =  price::call_price(call.get_debt(), call.get_collateral(),
-                                                   _bitasset_data->current_feed.maintenance_collateral_ratio);
-          }
+         call.collateral += o.delta_collateral.amount;
+         call.debt       += o.delta_debt.amount;
+         if( call.debt > 0 )
+         {
+            call.call_price  =  price::call_price(call.get_debt(), call.get_collateral(),
+                                                  _bitasset_data->current_feed.maintenance_collateral_ratio);
+         }
+         call.target_collateral_ratio = new_target_cr;
```

Comment:
I don't think `optional<optional<uint16_t>>` can be jsonified correctly.

Since type of that field is `optional` already, it would be always confusing about how to delete it, IMHO it's easier to treat it as an always-required field, aka always update.

Another approach is to use 2 fields, one for existence and the other for the real value. But I think it's too complicated since the `optional` type is just for doing this.

Reply:
I'd still prefer to keep the "if absent leave untouched" semantic. I suppose the option will not be available in the UI (all UI's) very soon, so someone who is used to trade via UI might want to set the tcr once via cli_wallet and continue normal operation using the UI.

Since tcr < 1000 doesn't make sense anyway, you could treat and invalid value (i. e. < 1000) as a deliberate delete request.

Reply:
I still think it's clearer to just update as is (which is written in BSIP).

Compare the two approaches:

1. update as is
* found null in op -> set null in obj
* found x in op -> set x in obj

2. update by checking if it's a special number
* found null in op -> don't change obj (no matter what it was)
* found 0 in op -> set null in obj
* found x(>0) in op -> set x(>0) in obj

IMHO the 2nd one will lead to more confusion.

Reply:
Another thing related is that `call_order_update_operation` is not only for updating a call order, but also for creating and removing, although removing is irrelevant here.

If the option is in "if absent leave untouched" semantic, the user may expect that the system will "remember" the option when a position is closed (either intendedly or unintendedly), also sometimes may forget to set when recreating a new position, both may lead to unintended behavior.

Compare the two approaches (I think the 1st one is better):

1. create new position as is
* found null in op -> set null in obj
* found x in op -> set x in obj

2. create new position checking if it's a special number
* found null in op -> set null in obj?
* found 0 in op -> set null in obj
* found x(>0) in op -> set x(>0) in obj

Reply:
Well, seems we have to agree to disagree (again ;-) ).

BSIP is authoritative though, so I'm OK with leaving it as it is.

Reply:
Perhaps propose a BSIP change? E.G. drop `optional` from object, use value `0` as "not set" instead.